File: sem-1/Programowanie-sieciowe/project/v_01/C/wireless_power_tool.c
/**
 * gcc -Wall wireless_power_tool.c -o wireless_power_tool
 * sudo ./wireless_power_tool
 * 
 * $ sudo ./wireless_power_tool
--- Narzędzie do sterowania mocą karty bezprzewodowej (C) ---

Dostępne interfejsy bezprzewodowe:
1. wlan0 (Aktualna moc: 20.00 dBm)

Wybierz interfejs (podaj numer): 1
Podaj nową moc nadawania dla 'wlan0' (w dBm): 15

Konfigurowanie interfejsu 'wlan0'...
1. Wyłączanie interfejsu 'wlan0'...
2. Ustawianie mocy nadawania na 15 dBm...
3. Włączanie interfejsu 'wlan0'...

Konfiguracja zakończona pomyślnie!

Sprawdzanie nowej konfiguracji...
Nowa moc dla interfejsu 'wlan0': 15.00 dBm
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h> // Dla geteuid()
#include <ctype.h>  // Dla isspace()

#define MAX_CMD_LEN 256      // Maksymalna długość polecenia
#define MAX_BUF_LEN 1024     // Maksymalna wielkość bufora dla odpowiedzi
#define MAX_INTERFACES 10    // Maksymalna liczba obsługiwanych interfejsów
#define MAX_IFACE_LEN 16     // Maksymalna długość nazwy interfejsu

// Struktura do przechowywania informacji o interfejsie
struct wireless_interface {
    char name[MAX_IFACE_LEN];
    char tx_power[32];
};

/**
 * @brief Wykonuje polecenie systemowe i zapisuje jego wyjście do bufora.
 *
 * @param command Polecenie do wykonania.
 * @param buffer Bufor, do którego zostanie zapisane wyjście.
 * @param buffer_size Rozmiar bufora.
 * @return int 0 w przypadku sukcesu, -1 w przypadku błędu.
 */
int execute_command_get_output(const char *command, char *buffer, size_t buffer_size) {
    FILE *fp;
    fp = popen(command, "r");
    if (fp == NULL) {
        perror("Błąd podczas uruchamiania polecenia (popen)");
        return -1;
    }

    size_t bytes_read = fread(buffer, 1, buffer_size - 1, fp);
    buffer[bytes_read] = '\0'; // Zapewnienie zakończenia stringa

    if (pclose(fp) == -1) {
        perror("Błąd podczas zamykania strumienia (pclose)");
        // Kontynuujemy, bo mogliśmy odczytać dane, nawet jeśli pclose zwróci błąd
    }
    return 0;
}

/**
 * @brief Wykonuje polecenie systemowe, które nie wymaga odczytu wyjścia.
 *
 * @param command Polecenie do wykonania.
 * @return int 0 w przypadku sukcesu, -1 w przypadku błędu.
 */
int execute_command_no_output(const char *command) {
    int status = system(command);
    if (status == -1 || WEXITSTATUS(status) != 0) {
        fprintf(stderr, "Błąd podczas wykonywania polecenia: %s\n", command);
        return -1;
    }
    return 0;
}

/**
 * @brief Parsuje wyjście polecenia 'iw dev' w poszukiwaniu nazw interfejsów.
 *
 * @param output Tekst do sparsowania.
 * @param interfaces Tablica struktur do wypełnienia.
 * @return int Liczba znalezionych interfejsów.
 */
int parse_interfaces(char *output, struct wireless_interface *interfaces) {
    int count = 0;
    char *line = strtok(output, "\n");
    while (line != NULL && count < MAX_INTERFACES) {
        char *iface_ptr = strstr(line, "Interface ");
        if (iface_ptr) {
            sscanf(iface_ptr + strlen("Interface "), "%15s", interfaces[count].name);
            count++;
        }
        line = strtok(NULL, "\n");
    }
    return count;
}

/**
 * @brief Pobiera i parsuje aktualną moc nadawania dla danego interfejsu.
 *
 * @param interface Struktura interfejsu do zaktualizowania.
 */
void get_current_tx_power(struct wireless_interface *interface) {
    char command[MAX_CMD_LEN];
    char buffer[MAX_BUF_LEN];
    snprintf(command, sizeof(command), "iw dev %s link", interface->name);

    strcpy(interface->tx_power, "Nieznana"); // Wartość domyślna

    if (execute_command_get_output(command, buffer, sizeof(buffer)) == 0) {
        char *tx_ptr = strstr(buffer, "txpower");
        if (tx_ptr) {
            // Przesuwamy wskaźnik za słowo "txpower" i białe znaki
            tx_ptr += strlen("txpower");
            while (*tx_ptr && isspace(*tx_ptr)) {
                tx_ptr++;
            }
            // Kopiujemy wartość mocy (np. "20.00 dBm")
            sscanf(tx_ptr, "%31[^\n]", interface->tx_power);
        }
    }
}


/**
 * @brief Ustawia nową moc nadawania dla wybranego interfejsu.
 *
 * @param interface_name Nazwa interfejsu.
 * @param power_dbm Moc w dBm.
 * @return int 0 w przypadku sukcesu, -1 w przypadku błędu.
 */
int set_tx_power(const char *interface_name, int power_dbm) {
    char command[MAX_CMD_LEN];

    printf("\nKonfigurowanie interfejsu '%s'...\n", interface_name);

    // 1. Wyłączenie interfejsu
    printf("1. Wyłączanie interfejsu '%s'...\n", interface_name);
    snprintf(command, sizeof(command), "ip link set %s down", interface_name);
    if (execute_command_no_output(command) != 0) return -1;

    // 2. Ustawienie nowej mocy
    printf("2. Ustawianie mocy nadawania na %d dBm...\n", power_dbm);
    snprintf(command, sizeof(command), "iw dev %s set txpower fixed %dmBm", interface_name, power_dbm);
    if (execute_command_no_output(command) != 0) {
        fprintf(stderr, "Nie udało się ustawić mocy. Próba przywrócenia interfejsu do stanu początkowego.\n");
        snprintf(command, sizeof(command), "ip link set %s up", interface_name);
        execute_command_no_output(command); // Spróbuj włączyć interfejs z powrotem
        return -1;
    }

    // 3. Włączenie interfejsu
    printf("3. Włączanie interfejsu '%s'...\n", interface_name);
    snprintf(command, sizeof(command), "ip link set %s up", interface_name);
    if (execute_command_no_output(command) != 0) return -1;

    printf("\nKonfiguracja zakończona pomyślnie!\n");
    return 0;
}


int main() {
    if (geteuid() != 0) {
        fprintf(stderr, "Ten program wymaga uprawnień administratora (root).\n");
        fprintf(stderr, "Uruchom go używając 'sudo ./twoj_program'\n");
        return 1;
    }

    printf("--- Narzędzie do sterowania mocą karty bezprzewodowej (C) ---\n");

    char buffer[MAX_BUF_LEN];
    if (execute_command_get_output("iw dev", buffer, sizeof(buffer)) != 0) {
        return 1;
    }

    struct wireless_interface interfaces[MAX_INTERFACES];
    int interface_count = parse_interfaces(buffer, interfaces);

    if (interface_count == 0) {
        printf("\nNie znaleziono żadnych interfejsów bezprzewodowych.\n");
        printf("Upewnij się, że karta Wi-Fi jest włączona i sterowniki są poprawnie zainstalowane.\n");
        return 1;
    }

    printf("\nDostępne interfejsy bezprzewodowe:\n");
    for (int i = 0; i < interface_count; i++) {
        get_current_tx_power(&interfaces[i]);
        printf("%d. %s (Aktualna moc: %s)\n", i + 1, interfaces[i].name, interfaces[i].tx_power);
    }

    int choice = 0;
    printf("\nWybierz interfejs (podaj numer): ");
    if (scanf("%d", &choice) != 1 || choice < 1 || choice > interface_count) {
        fprintf(stderr, "Nieprawidłowy wybór.\n");
        return 1;
    }
    const char *selected_interface_name = interfaces[choice - 1].name;

    int power = 0;
    printf("Podaj nową moc nadawania dla '%s' (w dBm): ", selected_interface_name);
    if (scanf("%d", &power) != 1) {
        fprintf(stderr, "Nieprawidłowa wartość. Proszę podać liczbę.\n");
        return 1;
    }

    if (set_tx_power(selected_interface_name, power) == 0) {
        printf("\nSprawdzanie nowej konfiguracji...\n");
        // Aktualizacja informacji o mocy po zmianie
        get_current_tx_power(&interfaces[choice - 1]);
        printf("Nowa moc dla interfejsu '%s': %s\n", interfaces[choice-1].name, interfaces[choice-1].tx_power);
    } else {
        fprintf(stderr, "\nWystąpił błąd podczas konfiguracji interfejsu.\n");
        return 1;
    }

    return 0;
}

File: sem-1/Programowanie-sieciowe/project/v_01/python/wireless_power_tool.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
'''
sudo python3 wireless_power_tool.py

$ sudo python3 wireless_power_tool.py
--- Narzędzie do sterowania mocą karty bezprzewodowej ---

Dostępne interfejsy bezprzewodowe:
1. wlan0 (Aktualna moc: 20.00 dBm)

Wybierz interfejs (podaj numer): 1
Podaj nową moc nadawania dla 'wlan0' (w dBm): 15

Konfigurowanie interfejsu 'wlan0'...
1. Wyłączanie interfejsu 'wlan0'...
2. Ustawianie mocy nadawania na 15 dBm...
3. Włączanie interfejsu 'wlan0'...

Konfiguracja zakończona pomyślnie!

Sprawdzanie nowej konfiguracji...
Nowa moc dla interfejsu 'wlan0': 15.00 dBm
'''
import subprocess
import sys
import re

def run_command(command):
    """
    Wykonuje polecenie systemowe i zwraca jego wynik.
    """
    try:
        process = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True, shell=True)
        stdout, stderr = process.communicate()
        if process.returncode != 0:
            print(f"Błąd podczas wykonywania polecenia: {command}")
            print(f"Stderr: {stderr.strip()}")
            return None
        return stdout.strip()
    except FileNotFoundError:
        print(f"Błąd: Polecenie '{command.split()[0]}' nie zostało znalezione. Upewnij się, że narzędzia 'iw' oraz 'ip' są zainstalowane.")
        sys.exit(1)
    except Exception as e:
        print(f"Wystąpił nieoczekiwany błąd: {e}")
        sys.exit(1)

def get_wireless_interfaces():
    """
    Pobiera listę dostępnych interfejsów bezprzewodowych.
    """
    output = run_command("iw dev")
    if not output:
        return []
    
    # Użycie wyrażenia regularnego do znalezienia nazw interfejsów
    interfaces = re.findall(r'Interface\s+(\w+)', output)
    return interfaces

def get_current_tx_power(interface):
    """
    Pobiera aktualną moc nadawania dla danego interfejsu.
    """
    output = run_command(f"iw dev {interface} link")
    if not output or "Not connected" in output:
        # Jeśli interfejs nie jest połączony, spróbuj innej komendy
        output = run_command(f"iw dev {interface} info")

    if not output:
        return "Nie można odczytać mocy"

    # Wyszukanie linii zawierającej 'txpower'
    match = re.search(r'txpower\s+([\d\.]+\s+dBm)', output)
    if match:
        return match.group(1)
    return "Nieznana"

def set_tx_power(interface, power_dbm):
    """
    Ustawia moc nadawania dla wybranego interfejsu.
    """
    print(f"\nKonfigurowanie interfejsu '{interface}'...")

    # 1. Wyłączenie interfejsu
    print(f"1. Wyłączanie interfejsu '{interface}'...")
    if run_command(f"sudo ip link set {interface} down") is None:
        return False

    # 2. Ustawienie nowej mocy
    print(f"2. Ustawianie mocy nadawania na {power_dbm} dBm...")
    # 'iw' oczekuje wartości w mBm (decybelomiliwatach), ale nowsze wersje przyjmują też dBm
    if run_command(f"sudo iw dev {interface} set txpower fixed {power_dbm}mBm") is None:
         # Próba alternatywnej komendy, jeśli pierwsza zawiedzie
        if run_command(f"sudo iw dev {interface} set txpower fixed {power_dbm}") is None:
            print("Nie udało się ustawić mocy. Próbuję przywrócić interfejs do stanu początkowego.")
            run_command(f"sudo ip link set {interface} up")
            return False

    # 3. Włączenie interfejsu
    print(f"3. Włączanie interfejsu '{interface}'...")
    if run_command(f"sudo ip link set {interface} up") is None:
        return False
    
    print("\nKonfiguracja zakończona pomyślnie!")
    return True


def main():
    """
    Główna funkcja programu.
    """
    if os.geteuid() != 0:
        print("Ten skrypt wymaga uprawnień administratora (root) do zmiany ustawień karty sieciowej.")
        print("Uruchom go używając 'sudo python3 wireless_power_tool.py'")
        sys.exit(1)

    print("--- Narzędzie do sterowania mocą karty bezprzewodowej ---")
    
    interfaces = get_wireless_interfaces()
    if not interfaces:
        print("\nNie znaleziono żadnych interfejsów bezprzewodowych.")
        print("Upewnij się, że karta Wi-Fi jest włączona i sterowniki są poprawnie zainstalowane.")
        sys.exit(1)

    print("\nDostępne interfejsy bezprzewodowe:")
    for i, iface in enumerate(interfaces, 1):
        power = get_current_tx_power(iface)
        print(f"{i}. {iface} (Aktualna moc: {power})")

    try:
        choice = int(input("\nWybierz interfejs (podaj numer): "))
        if not 1 <= choice <= len(interfaces):
            print("Nieprawidłowy wybór.")
            sys.exit(1)
        selected_interface = interfaces[choice - 1]

        power = int(input(f"Podaj nową moc nadawania dla '{selected_interface}' (w dBm): "))
        
        set_tx_power(selected_interface, power)

        print("\nSprawdzanie nowej konfiguracji...")
        new_power = get_current_tx_power(selected_interface)
        print(f"Nowa moc dla interfejsu '{selected_interface}': {new_power}")

    except ValueError:
        print("Nieprawidłowa wartość. Proszę podać liczbę.")
    except KeyboardInterrupt:
        print("\nPrzerwano działanie programu.")
        sys.exit(0)

if __name__ == "__main__":
    import os
    main()